### ISSUE TITLE
Dispose of PyObject when finalized.

### ISSUE THREAD

**Issue by PuerNoctis** [pythonnet/pythonnet#1045]
### Environment

-   Pythonnet version: 2.4.1
-   Python version: 3.7
-   Operating System: Windows 10

### Details

While working with Keras.NET on one of our training projects, we found that our application crashed with a heap corruption somewhere within Python. We then stared to debug into the Python.NET code and found, that, for whatever reason, the Dispose() of the PyObject was called after the class was already finalized.

During this, the reference counter was decreased and the already freed pointer was deallocated again. We don't know when and where (probably Keras) is finalizing the objects, but we assume that the Dispose in PyObject should check for finalization in any case.

The relevant code is in the 'runtime/PyObject.cs' Dispose() method. We made the simple change to just check the '**_finalized**' flag also:

```csharp
protected virtual void Dispose(bool disposing)
 {
     if (!disposed)
     {
         if (Runtime.Py_IsInitialized() > 0 && !Runtime.IsFinalizing && !_finalized)
         {
             IntPtr gs = PythonEngine.AcquireLock();
             Runtime.XDecref(obj);
             obj = IntPtr.Zero;
             PythonEngine.ReleaseLock(gs);
         }
         disposed = true;
     }
}
```

Does anyone see issues with this change? If not, I would prepare a pull request.

Also, I am not too deep yet into the Python.NET code to gauge whether it is plausible to set 'obj' to IntPtr.Zero in the class' finalizer. Maybe that suffices also.

**Comment by lostmsu** [comment]
@PuerNoctis note, that Keras.NET is using a fork of this project: https://github.com/henon/pythonnet_netstandard

**Comment by lostmsu** [comment]
@PuerNoctis since you are holding a reference to the object, it should have not been delallocated/freed until you called `Dispose`, or until our finalizer gets to it.

Both `Finalizer`'s call to `Dispose` and your own one should have set `disposed` and thus prevented double decref.

Unless I am missing something, this should not be a problem, unless you also use multithreading, as calls to `PyObject.Dispose` are not thread safe.

@amos402 , can you look at this issue?

**Comment by amos402** [comment]
Dispose is thread safe. It may cause by some reference count error on your code or pythonnet internal. You can enable `FINALIZER_CHECK` for a runtime check.

**Comment by lostmsu** [comment]
@amos402 `Dispose` above is not thread safe, because `disposed` and `_finalized` are not checked inside any locks.

**Comment by amos402** [comment]
> 
> 
> @amos402 `Dispose` above is not thread safe, because checks for `disposed` and `_finalized` are not protected by any locks or atomics.

Indeed. But in this case just simply change into
```c#
protected virtual void Dispose(bool disposing)
        {
            if (disposed
                || Runtime.Py_IsInitialized() == 0
                || Runtime.IsFinalizing)
            {
                return;
            }
            disposed = true;
            IntPtr gs = PythonEngine.AcquireLock();
            try
            {
                if (disposed)
                {
                    return;
                }
                Runtime.XDecref(obj);
                obj = IntPtr.Zero;
            }
            finally
            {
                PythonEngine.ReleaseLock(gs);
            }
        }
```
This may enough, no `_finalized` needed(that's the flag for the deconstructor), no clr lock either.

**Comment by amos402** [comment]
> 
> 
> @amos402 `Dispose` above is not thread safe, because checks for `disposed` and `_finalized` are not protected by any locks or atomics.

Indeed. But in this case just simply change into
```c#
        protected virtual void Dispose(bool disposing)
        {
            if (disposed
                || Runtime.Py_IsInitialized() == 0
                || Runtime.IsFinalizing)
            {
                return;
            }
            disposed = true;
            IntPtr gs = PythonEngine.AcquireLock();
            try
            {
                if (obj == IntPtr.Zero)
                {
                    return;
                }
                IntPtr handle = obj;
                obj = IntPtr.Zero;
                Runtime.XDecref(handle);
            }
            finally
            {
                PythonEngine.ReleaseLock(gs);
            }
        }
```
This may enough, no `_finalized` needed(that's the flag for the deconstructor), no clr lock either.

**Comment by lostmsu** [comment]
@PuerNoctis can you provide more details? It is really unclear why the object, that has not been finalized yet is being collected?

**Comment by PuerNoctis** [comment]
Ok, so I did some more analyzing. First, after turning on the `FINALIZER_CHECK` I indeed got `IncorrectRefCountException`s just before or after the heap corruption crashed the `Finalizer` thread. The frames of the main thread showed that, at this point, Keras was waiting for what I assume was a native/python interop thread. As you suggested, there seems to be a lot of concurrency going on in the background (due to Keras).

After that, I turned on the `TRACE_ALLOC` feature to find out where the object in question was created at. It is a `PyDict` and, in some cases, one of the values it holds. These objects are created within Keras.NET just before invoking a python method to pass the arguments.

I assume that after the call is done, the GC calls the class finalizer on `PyDict` at some point and pushes it into the `Finalizer` queue. When the `Finalizer` then disposes the `PyDict` (or sometimes one of its containing objects) the corruption occurs.

I know that this is Keras.NET code, and I will update my opened issue at their repo accordingly, but just to show how the breaking code looks like:

```csharp
public PyObject InvokeMethod(string method, Dictionary<string, object> args)
{
   var pyargs = ToTuple(new object[]
   {
        args.FirstOrDefault().Value
   });

    var kwargs = new PyDict();

    bool skip = true;
    foreach (var item in args)
    {
        if (skip)
        {
            skip = false;
            continue;
        }

        if (item.Value != null && !string.IsNullOrWhiteSpace(item.Value.ToString()))
        {
            kwargs[item.Key] = ToPython(item.Value);
        }
    }

    if (args.Count > 0)
        return PyInstance.InvokeMethod(method, pyargs, kwargs);
    else
        return PyInstance.InvokeMethod(method, null, null);
}
```
Here the `kwargs` object is the problem. For testing, I kept references to that object alive (in order to prevent the class finalizer to run), and then there are no more heap corruptions. If I keep the reference alive however, and dispose the object manually after the call to `PyInstance.InvokeMethod`, I still get the same error again with `_finalized` being `false`.

Somewhere within that python call there seems to be a problem with the reference counting I assume.

This is how far my investigation has currently gotten. I think I'll just give Keras.NET an update and see if they can pinpoint the problem more clearly.

As for this issue, I think it can be closed unless you think there might be more to it than just wrong reference counting in Keras.NET and downstream. Oh, and as a side-note: My initial suggestion to check for the `_finalized` flag is even dangerous I found, because when the `Finalizer` runs, all the objects have that flag set to `true`, so no resources got freed in the end and memory fills up quite quickly. Could have figured that one out earlier, sorry.

**Comment by lostmsu** [comment]
@PuerNoctis thanks! I am going to close this issue, as from your last example it seems like the object is decref'ed in `PyInstance.InvokeMethod` unnecessarily.

**Comment by MirunaTirca** [comment]
We seem to encounter the same problem with our project: we are doing a migration to python 3.6 and switched to pythonnet 2.4.0, and we get the following crash:

`Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at Python.Runtime.Runtime.PyErr_Fetch(IntPtr& ob, IntPtr& val, IntPtr& tb)
   at Python.Runtime.PyObject.Dispose(Boolean disposing)
   at Python.Runtime.PyObject.Dispose()
   at Python.Runtime.Finalizer.DisposeAll()
   at Python.Runtime.Finalizer.ThrottledCollect()
   at Python.Runtime.PyTuple..ctor()
   .... some calls from our project ....
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)`
   
All logs seem to indicate a corruption of memory but can't point the exact cause for it.
Considering this thread and also the Keras.Net issue [(),](https://github.com/SciSharp/Keras.NET/issues/94) can you please confirm if the issue is from pythonnet side or the project side - here I am refering to the closing argument in Keras.Net: "I guess this issue is fixed already with pythonnet. Let me know if still have issues".

For the Keras.Net reported issue back in 2020, what exactly was the solution?
If the issue in the Keras.Net case was in pythonnet, can you please provide some guidance?
Also, do you think it would be better to upgrade to a newer pythonnet? Can you confirm if that would solve the crash?

**Comment by lostmsu** [comment]
pythonnet 2.4.0 is not really supported anymore (and neither is Python 3.6)

**Comment by MirunaTirca** [comment]
Unfortunately, at this time, another upgrade is costly.
Do you know how the issue was solved in 2020 for Keras.Net?

**Comment by lostmsu** [comment]
They are simply saying that we (e.g. Python.NET maintainers) fixed it. And yes, the fix is in pythonnet 3.0
